File: Backtracking_and_Search/LC_90_Subsets_II.py
# Written by: Jian-Yuan Lin
# Date: 2022.11.19

class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        ans = []
        n = len(nums)

        def dfs(l, nums, start, curr, ans):
            if len(curr) == l:
                ans.append(curr[:])
                return
            for i in range(start, n):
                if i != start and nums[i] == nums[i-1]:
                    continue
                curr.append(nums[i])
                dfs(l, nums, i+1, curr, ans)
                curr.pop()

        for l in range(n+1):
            dfs(l, nums, 0, [], ans)
        return ans

File: Backtracking_and_Search/LC_90_Subsets_II.cpp
// Written by: Jian-Yuan Lin
// Date: 2022.11.19

class Solution {
public:
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> ans;
        vector<int> curr;
        int n = nums.size();
        for (int l = 0; l <= n; l++){
            dfs(nums, curr, 0, l, ans);
        }
        return ans;
    }
private:
    void dfs(vector<int>& nums, vector<int>& curr, int start, int l, vector<vector<int>>& ans){
        if (curr.size() == l){
            ans.push_back(curr);
            return;
        }
        for(int i = start; i < nums.size(); i++){
            if (i != start && nums[i] == nums[i-1]){
                continue;
            }
            curr.push_back(nums[i]);
            dfs(nums, curr, i+1, l, ans);
            curr.pop_back();
        }
    }
};